src/v/kafka/data/rpc/service.cc

```diff
@@ -175,6 +177,20 @@ ss::future<result<model::offset, cluster::errc>> local_service::produce(
     if (!topic_cfg) {
         co_return cluster::errc::topic_not_exists;
     }
+    if constexpr (std::is_same_v<decltype(ntp), const model::ntp&>) {
+        if (!_shadow_link_registry->is_topic_mutable(ntp.tp.topic))
+          [[unlikely]] {
+            co_return cluster::errc::partition_operation_failed;
+        }
+    } else if constexpr (std::derived_from<decltype(ntp), model::ktp>) {
+        if (!_shadow_link_registry->is_topic_mutable(ntp.get_topic()))
+          [[unlikely]] {
+            co_return cluster::errc::partition_operation_failed;
+        }
+    } else {
+        static_assert(false, "ntp must be model::ntp or model::ktp");
```

Comment:
@michael-redpanda does this need to be a type-dependent false? that was my understanding, but it seems this is compiling. so i'm not sure..

Reply:
This made it into c++23

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2593r1.html